src/libraries/System.Threading.RateLimiting/src/System/Threading/RateLimiting/PartitionedRateLimiter.cs

```diff
@@ -0,0 +1,243 @@
+        /// </summary>
+        /// <typeparam name="TResource">The resource type that is being rate limited.</typeparam>
+        /// <typeparam name="TPartitionKey">The type to distinguish partitions with.</typeparam>
+        /// <param name="partitioner">Method called every time an Acquire or WaitAsync call is made to figure out what rate limiter to apply to the request.
+        /// If the <see cref="RateLimitPartition{TKey}.PartitionKey"/> matches a cached entry then the rate limiter previously used for that key is used. Otherwise, the factory is called to get a new rate limiter.</param>
+        /// <param name="equalityComparer">Optional <see cref="IEqualityComparer{T}"/> to customize the comparison logic for <typeparamref name="TPartitionKey"/>.</param>
+        /// <returns></returns>
+        public static PartitionedRateLimiter<TResource> Create<TResource, TPartitionKey>(
+            Func<TResource, RateLimitPartition<TPartitionKey>> partitioner,
+            IEqualityComparer<TPartitionKey>? equalityComparer = null) where TPartitionKey : notnull
+        {
+            return new DefaultPartitionedRateLimiter<TResource, TPartitionKey>(partitioner, equalityComparer);
+        }
+    }
+
+    internal sealed class DefaultPartitionedRateLimiter<TResource, TKey> : PartitionedRateLimiter<TResource> where TKey : notnull
+    {
+        private readonly Func<TResource, RateLimitPartition<TKey>> _partitioner;
+
+        // TODO: Look at ConcurrentDictionary to try and avoid a global lock
+        private Dictionary<TKey, Lazy<RateLimiter>> _limiters;
+        private bool _disposed;
+
+        // Used by the Timer to call TryRelenish on ReplenishingRateLimiters
+        // We use a separate list to avoid running TryReplenish (which might be user code) inside our lock
+        // And we cache the list to amortize the allocation cost to as close to 0 as we can get
+        private List<Lazy<RateLimiter>>? _cachedLimiters = new();
+        private bool _cacheInvalid;
+        private TimerAwaitable _timer;
+        private Task _timerTask;
+
+        // Use the Dictionary as the lock field so we don't need to allocate another object for a lock and have another field in the object
+        private object Lock => _limiters;
+
+        public DefaultPartitionedRateLimiter(Func<TResource, RateLimitPartition<TKey>> partitioner,
+            IEqualityComparer<TKey>? equalityComparer = null)
+        {
+            _limiters = new Dictionary<TKey, Lazy<RateLimiter>>(equalityComparer);
+            _partitioner = partitioner;
+
+            // TODO: Figure out what interval we should use
+            _timer = new TimerAwaitable(TimeSpan.FromMilliseconds(100), TimeSpan.FromMilliseconds(100));
+            _timerTask = RunTimer();
+        }
+
+        private async Task RunTimer()
+        {
+            _timer.Start();
+            while (await _timer)
+            {
+                Replenish(this);
```

Comment:
I think we're going to have to catch any exceptions now that we rely on the loop continuing. It's unfortunate that we cannot log it this stops any exception from showing up as a `UnhandledException`. It's tempting to use `ExceptionDispatchInfo.Capture` and dispatch the throwing of the exception to a threadpool thread.